### ISSUE TITLE
UTF-8 URL encoded value

### ISSUE THREAD

**Issue by BHMath** [divolte/divolte-collector#98]
I noticed that accented characters from url are not well mapped in my avro data.

Here my mapping file
```
  map location() onto 'location'
  def locationUri = parse location() to uri
  map locationUri.decodedFragment() onto 'localPath'
```

And here some url and the result
http://localhost:8080/#Détails
```
  "location" : "http://127.0.0.1:8180/myapp/#DÃ©tails",
  "localPath" : "DÃ©tails",
```
http://localhost:8080/#D%C3%A9tails
```
  "location" : "http://127.0.0.1:8180/myapp/#D%C3%A9tails",
  "localPath" : "DÃ©tails",
```

And I also noticed that if the fragment contains any spaces my local path wont register
http://localhost:8080/#Mes Infos
```
  "location" : "http://127.0.0.1:8180/myapp/#Mes Infos",
  "localPath" : "",
```

**Comment by friso** [comment]
Could you provide a more complete description of how you got these result? What sits between the browser and Divolte? What are you using to view the resulting data (shell? text editor?). I have a hard time reproducing this. I haven't tried the mapping of the decodedFragment with your example yet, but with all defaults, I get this:

```
  "location": {
    "string": "http://localhost:8290/#Café"
  }
```

or this:
```
  "location": {
    "string": "http://localhost:8290/#D%C3%A9tails"
  }
```
Both of these should decode correctly.

**Comment by BHMath** [comment]
My webapp is a Tapestry application running inside a tomcat, I access it trhu Google Chrome. 
Divolte is installed on a CentOS VM and I just use the avro-tools provided by divolte to visualize the data.

**Comment by friso** [comment]
In the shell where you run avro-tools, can you try this and check the output: `echo -e "\xc3\xa9"`? It should print é.

**Comment by jbnicolai** [comment]
> And I also noticed that if the fragment contains any spaces my local path wont register
http://localhost:8080/#Mes Infos

See the second section of 2.4.3 in the IETF's URI spec https://www.ietf.org/rfc/rfc2396.txt and the `unsafe` section of 2.2 in the URL spec  https://www.ietf.org/rfc/rfc1738.txt

**Comment by jbnicolai** [comment]
@BHMath note that what you're using is not URLs or URIs but a [IRI](https://en.wikipedia.org/wiki/Internationalized_resource_identifier). This is **not** supported in HTML4 and older browsers. See [this stackoverflow question](https://stackoverflow.com/questions/2742852/unicode-characters-in-urls) for a more thorough explanation.

tl;dr: use percent encoding, not raw unicode characters.

**Comment by jbnicolai** [comment]
@friso I somewhat doubt that's the issue, but in case it is: try `echo $LANG`and make sure it's not `C`. If it is, switch is to something like `en_US`

**Comment by BHMath** [comment]
Here the result of the two commands

```
[www@localhost ~]$ echo -e "\xc3\xa9"
Ã©
[www@localhost ~]$ echo $LANG
en_US.UTF-8
[www@localhost ~]$
```
I don't really understand what I can do about my space. Currently I replace it with dashes and everything is ok and I'm ok with it.

**Comment by friso** [comment]
Apart from the `LANG` env var, the terminal settings are also important. I'm not a Ubuntu user, but I imagine that there's somewhere you can set the encoding for the terminal display itself.

Instead of displaying the result in your terminal, try redirecting the output to a file and then opening that file in a text editor; e.g.: `avro-tools tojson your-file.avro > /tmp/your-file.json`.

As for spaces (and accented characters), you should use percent encoding in URIs, really. That means when rendering the page, you need to make sure all href's are percent encoded. Of course, if you want the URIs to be human readable, you'll want to replace special characters by readable ones (so spaces can become - and accented characters can be ascii-folded).

**Comment by BHMath** [comment]
ok thanks... The problem is my putty. I did save to a txt file, and get it to my windows platform. When I open myfile.txt with notepad++ the accented characters are well displayed. Thanks.

About my URL, I look into the developper console to really see what is sent to divolte.

If my browser url is
```
http://localhost:8180/myapp/#my Infos
```
then the get sent to divolte is 
```
http://localhost:8290/csc-event?p=0%3Ai8u8qdqd%3A6VMPpVYW3Jiw79CTf0lqslV3Rbc6~Egi&s=0%3Aiafeth2t%3AWPYBJgoEvtOrCGhHKUfUqfHm7XuOwTK2&v=0%3ANdUOozDeZySFpYsWex462w_~7HiCfHN~&e=0%3ANdUOozDeZySFpYsWex462w_~7HiCfHN~0&c=iafeu55d&n=f&f=f&l=http%3A%2F%2Flocalhost%3A8180%2Fmyapp%2F%23my%20Infos&i=11y&j=lc&k=1&w=11y&h=iz&t=displayPage&u=(suserId!!)&x=-wtx5ut
```
So what the info sent to divolte are percent encoding right ? What did I miss ?

**Comment by friso** [comment]
Yeah, so the divolte javascript sends the location with a space character as percent encoded query parameter to our server side (it takes `window.href.location` and encodes it using `encodeURIComponent(...)`), where it is decoded again into a `String` with a space in it on the server side. This is subsequently parsed using `java.net.URI`, which per [the docs](https://docs.oracle.com/javase/8/docs/api/java/net/URI.html) has its own way of handling spaces in a URI. Here's a simple test in a Groovy shell:
```
groovy:000> u = new URI("http://localhost:8290/#Café au lait")
ERROR java.net.URISyntaxException:
Illegal character in fragment at index 27: http://localhost:8290/#Café au lait
```

So URIs with spaces result in an exception being thrown by the parser, which in turn results in the value being absent as far as the mapping is concerned; as a result you get default values in the record. Our only option to deviate from this would be to write our own URI parsing, which I think is overkill. We should probably note in the docs that we don't parse URIs with spaces well, but we'd have to test the corner cases first, to figure out how best to express the actual constraints.

**Comment by BHMath** [comment]
Ok, that's cool with me. I know that spaces aren't handle so i'll try to not use them/
